pkg/backends/influxdb/flush.go

```diff
@@ -0,0 +1,256 @@
+//
+// Process unnamed, sort values, and concatenate to a single value:
+// unnamed: ["foo", "baz"] -> sort as ["baz", "foo"] -> concatenate as "unnamed=baz__foo"
+// key: ["bar", "thing"] -> sort as ["bar", "thing"] -> concatenate as "key=bar__thing"
+// other: ["something"] -> sort as ["something"] -> concatenate as "other=something"
+//
+// Sort by key and merge:
+// key=bar__thing,other=something,unnamed=baz__foo
+//
+// The name will be prepended, and a space will be appended.
+func formatNameTags(name string, tags gostatsd.Tags) string {
+	newTags := make(map[string][]string, len(tags))
+	for _, tag := range tags {
+		kv := strings.SplitN(tag, ":", 2)
+
+		var key, value string
+		if len(kv) == 1 {
+			key = "unnamed"
+			value = kv[0]
+		} else {
+			key = kv[0]
+			value = kv[1]
+		}
+		newTags[key] = append(newTags[key], value)
+	}
+
+	sb := &strings.Builder{}
+	escapeNameToBuilder(sb, name)
+	keys := make([]string, 0, len(newTags))
+	for key, values := range newTags {
+		keys = append(keys, key)
+		sort.Strings(values)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		sb.WriteByte(',')
+		escapeTagToBuilder(sb, key)
+		sb.WriteByte('=')
+		for idx, value := range newTags[key] {
+			if idx > 0 {
+				sb.WriteString("__")
+			}
+			escapeTagToBuilder(sb, value)
+		}
+	}
+	sb.WriteByte(' ')
+	return sb.String()
+}
+
+// escape will escape NL (\n), CR (\r), TAB (\t), space, comma, backslash, and equals
+// writing the results to the supplied strings.Builder.
+func escapeTagToBuilder(sb *strings.Builder, s string) {
+	for _, ch := range s {
+		if ch == '\n' {
+			sb.WriteString("\\n")
+		} else if ch == '\r' {
+			sb.WriteString("\\r")
+		} else if ch == '\t' {
+			sb.WriteString("\\t")
+		} else {
+			if ch == ' ' || ch == ',' || ch == '\\' || ch == '=' {
+				sb.WriteByte('\\')
+			}
+			sb.WriteRune(ch)
+		}
+	}
+}
+
+// escape will escape NL (\n), CR (\r), TAB (\t), space, comma, and backslash,
+// writing the results to the supplied strings.Builder.
+func escapeNameToBuilder(sb *strings.Builder, s string) {
+	for _, ch := range s {
+		if ch == '\n' {
+			sb.WriteString("\\n")
+		} else if ch == '\r' {
+			sb.WriteString("\\r")
+		} else if ch == '\t' {
+			sb.WriteString("\\t")
+		} else {
+			if ch == ' ' || ch == ',' || ch == '\\' {
+				sb.WriteByte('\\')
+			}
+			sb.WriteRune(ch)
+		}
+	}
+}
+
+// escapeStringToBuilder will escape double quotes and backslash writing the
+// results to the supplied strings.Builder, surrounded by double-quotes
+func escapeStringToBuilder(sb *strings.Builder, s string) {
+	sb.WriteByte('"')
+	for _, ch := range s {
+		if ch == '\\' || ch == '"' {
+			sb.WriteByte('\\')
+		}
+		sb.WriteRune(ch)
+	}
+	sb.WriteByte('"')
+}
+
+// writeName writes "measurement,tag1=value1,tag2=value2,... " to the output buffer.
+// Common to all metric types and events.
+func writeName(w io.Writer, name string, tags gostatsd.Tags) {
+	_, _ = w.Write([]byte(formatNameTags(name, tags)))
```

Comment:
Is discarding the write error worth while here? 
I worry this could bite us later.

Reply:
Yep, errors are propagated out later when we flush/close.  The gzip writer has a "sticky" error.

https://blog.golang.org/errors-are-values